src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs

```diff
@@ -113,41 +120,34 @@ private void SetEntry(CacheEntry entry)
             // Initialize the last access timestamp at the time the entry is added
             entry.LastAccessed = utcNow;
 
-            var added = false;
             CacheEntry priorEntry;
-
-            _entryLock.EnterWriteLock();
-            try
+            if (_entries.TryRemove(entry.Key, out priorEntry))
```

Comment:
I agree that readers shouldn't get `null` if a writer is updating.  If a reader and writer are racing, of course it's fine for the reader to get the old or new value.  What is the observable difference between `ReaderWriterLock` and the new change to `ConcurrentDictionary`?